### PR TITLE
[5.2] SEF: Enforce suffix

### PR DESCRIPTION
**Pull request by Hackwar**
Pull Request for Issue #15025 .

### Summary of Changes
Joomla has been improving its SEO performance constantly and one issue which is still open is the behavior of the suffix. Right now, you can access a URL in Joomla with or without suffix when the option is enabled in the global configuration. This PR introduces a new setting the SEF system plugin which enforces a consistent suffix behavior.

When SEF URLs are enabled, the suffix is enabled and this option is enabled, Joomla will always redirect GET requests to a URL with a suffix if none is present. It will also redirect URLs with a query `format` parameter to the nicer URL and replace the "nice" suffix with the `format` parameter if the two collide.

With Joomla 6.0 the option to switch this on/off should be removed again and this should be the default behavior, which would then be added to `SiteRouter::parseFormat()`. Right now this is YASO (Yet Another Stupid Option) to allow people to test this in live systems and to switch it off if we encounter unforseen issues. The time from 5.1 to 6.0 could be seen as a grace period.

~~This PR depends on #42692.~~

I'd like to thank [djumla GmbH](https://www.djumla.de) for sponsoring this feature.

### Testing Instructions
1. Make sure that you have PR #42692 in your test installation
2. Apply the patch
3. Enable SEF URLs and the "Add Suffix to URL" options in global configuration
4. Go to "System - SEF" plugin and enable the option to "Enforce Suffix"
5. Open a URL in the frontend and see that URLs ending in a `/` don't have a suffix, everything else has a suffix.
6. Take a URL with a suffix and remove the suffix. See that Joomla redirects back to a URL with `.html`.
7. Take a URL with a suffix and add `?format=feed` to it, for example for a category blog view. See that you get the feed output with the right URL.

### Link to documentations
Please select:
- [X] Documentation link for docs.joomla.org: https://docs.joomla.org/Search_Engine_Friendly_URLs
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [X] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by brianteeman** [comment]
This looks like it addresses a 7 year old bug #15025

**Comment by ceford** [comment]
I have tested this item :white_check_mark: successfully on 784cfdae7d1732ea9ac9f40e699b9c8e36ad82ab<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by nielsnuebel** [comment]
I have tested this item :white_check_mark: successfully on 784cfdae7d1732ea9ac9f40e699b9c8e36ad82ab<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on 784cfdae7d1732ea9ac9f40e699b9c8e36ad82ab<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by brianteeman** [comment]
In favour of the bug fix. Not in favour of the option. As such this should be in 4.4 as a bug fix and not as a new feature in a future release

**Comment by Hackwar** [comment]
Link to documentation has been added.

**Comment by Hackwar** [comment]
This is a change in behavior and not just a bugfix. That's why we wont get around this option. For Joomla 6.0 this option will be removed again and the behavior added to the normal behavior of the `Add Suffix to URL` option. So as long as we will stick to semantic versioning we will have to add this as an option.

**Comment by Hackwar** [comment]
@Nuyonuyonoina could you comment what is failing on your end? Otherwise I can't fix that and this PR wont be merged.

**Comment by webfeuerflo** [comment]
failed. followed the instructions, but if I remove .html from the URL it does not get added back

**Comment by webfeuerflo** [comment]
I have tested this item :red_circle: unsuccessfully on 784cfdae7d1732ea9ac9f40e699b9c8e36ad82ab<br /><br />failed. followed the instructions, but if I remove .html from the URL it does not get added back<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by webfeuerflo** [comment]
Tested on PHP 8.1.27, 10.4.32-MariaDB, Joomla! 5.1.0-alpha4 Alpha [ Kudumisha ] 20-February-2024 16:31 GMT<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by webfeuerflo** [comment]
the pages can now be accessed with 3 URLs: with .html with / and without /. All show the same article
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by Hackwar** [comment]
Did you apply the PR #42692 as well?

**Comment by webfeuerflo** [comment]
yes, both patches are installed and active in the patchtester

**Comment by webfeuerflo** [comment]
Found out, why it failed! I had yootheme installed, although not set as standard template, the framework is doing something that interferes with this patch. When I deactivate all yootheme extensions, the patch works
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by webfeuerflo** [comment]
@saschadube the yootheme system plugin is keeping this patch from working. Any idea what is going on?
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by Hackwar** [comment]
My assumption would be that they are creating the router object so early in the process that no other plugin is able to run and thus these new options can't attach themselfs to the router. To be honest, this is something Yootheme needs to fix.

**Comment by richard67** [comment]
I've unset @webfeuerflo 's negative test result as it has turned out that it was not caused by this PR but by a 3rd party plugin.

**Comment by janschoenherr** [comment]
Thanks, the next YOOtheme Pro version will resolve the problem with accessing com_media and by that creating the SiteRouter early. 🙏

**Comment by Hackwar** [comment]
I'm sorry, but in that case I rather pull this PR back and not have this in 5.1. I'm not going to take the chances of sites breaking because this is enabled by default. There are way to many sites out there which have weird code to just take the chance on this.

**Comment by bembelimen** [comment]
Revert as it's a draft.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by Hackwar** [comment]
In the CMS maintainer meeting we talked about this again and the team agreed to the current implementation. Since it is to late for 5.1, I rebased this to 5.2, but 5.2 is in a desperate need of an upmerge... Anyway, this is theoretically ready to test.

File: plugins/system/sef/src/Extension/Sef.php [hunk 1]
```diff
@@ -268,6 +273,46 @@ function ($match) use ($base, $protocols) {
...
+
+        if (substr($route, -9) === 'index.php' || substr($route, -1) === '/') {
+            // We don't want suffixes when the URL ends in index.php or with a /
+            return;
+        }
+
+        $suffix       = pathinfo($route, PATHINFO_EXTENSION);
+        $nonSEFSuffix = $origUri->getVar('format');
+
+        if ($nonSEFSuffix && $suffix !== $nonSEFSuffix) {
```
Comment (SniperSister) [inline plugins/system/sef/src/Extension/Sef.php:307]: This causes issues with URLs following this pattern  ```http://localhost/content-component/article-category-blog?format=html``` as strlen($suffix) will be 0 and the substring returns and empty string.

The following code block fixes the issue:

```php
            // There is a URL query parameter named "format", which isn't the same to the suffix
            $pathWithoutSuffix = ($suffix !== '') ? substr($route, 0, -(\strlen($suffix) + 1)) : $route;

            $origUri->delVar('format');
            $origUri->setPath($pathWithoutSuffix . "." . $nonSEFSuffix);
            $this->getApplication()->redirect($origUri->toString(), 301);
```

**Comment by Hackwar** [comment]
Thank you, @SniperSister, I fixed that. @ceford, @nielsnuebel, @viocassel and of course anybody else who is intersted, would you be so kind to test this again, so that we can merge this for 5.2?

**Comment by ceford** [comment]
I have tested this item :white_check_mark: successfully on 5900b39ef1985e6b5c29dd7befb62c6e746c00c6<br /><br />I see #42692 has been merged. I am using a cms clone - so I did a git pull them npm ci, applied the patch and I get this:

An error has occurred.

    0 Call to undefined method Joomla\Plugin\System\Sef\Extension\Sef::setSiteRouter() 

I needed to disable the sef plugin in the database to get admin working again.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by ceford** [comment]
I have tested this item :red_circle: unsuccessfully on 5900b39ef1985e6b5c29dd7befb62c6e746c00c6<br /><br />I see #42692 has been merged. I am using a cms clone - so I did a git pull them npm ci, applied the patch and I get this:

An error has occurred.

    0 Call to undefined method Joomla\Plugin\System\Sef\Extension\Sef::setSiteRouter() 

I needed to disable the sef plugin in the database to get admin working again.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by Hackwar** [comment]
@ceford The comment about #42692 actually isn't true anymore and that PR actually has been reverted before the release of 5.1. Since this PR also doesn't touch that code, your git clone most likely is very outdated and should be updated to latest 5.2-dev before testing this.

**Comment by ceford** [comment]
My clone is up to date. I checked out 5.2-dev - title bar shows ‎5.2.0-alpha1-dev

The SEF plugin is enabled. I apply the patch and then all menu items in the back end give:

```
     0 Call to undefined method Joomla\Plugin\System\Sef\Extension\Sef::setSiteRouter() 

Call Stack
# 	Function 	Location
1 	() 	JROOT/plugins/system/sef/services/provider.php:42
2 	Joomla\DI\ServiceProviderInterface@anonymous/Users/ceford/Sites/joomla-cms5/plugins/system/sef/services/provider.php:22$97->{closure}() 	JROOT/libraries/vendor/joomla/di/src/ContainerResource.php:172
3 	Joomla\DI\ContainerResource->getInstance() 	JROOT/libraries/vendor/joomla/di/src/Container.php:95
4 	Joomla\DI\Container->get() 	JROOT/libraries/src/Extension/ExtensionManagerTrait.php:177
5 	Joomla\CMS\Application\CMSApplication->loadExtension() 	JROOT/libraries/src/Extension/ExtensionManagerTrait.php:99
6 	Joomla\CMS\Application\CMSApplication->bootPlugin() 	JROOT/libraries/src/Plugin/PluginHelper.php:232
7 	Joomla\CMS\Plugin\PluginHelper::import() 	JROOT/libraries/src/Plugin/PluginHelper.php:192
8 	Joomla\CMS\Plugin\PluginHelper::importPlugin() 	JROOT/libraries/src/Application/CMSApplication.php:765 
```
So I have to disable the plugin so I can revert the patch. It may be my installation that is the problem but I don't know what to do about it.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by SniperSister** [comment]
I have tested this item :white_check_mark: successfully on ef8201e7c63a4d0aee775aeaee33de63c757e9f1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by webnet-assmann** [comment]
I have tested this item :white_check_mark: successfully on ef8201e7c63a4d0aee775aeaee33de63c757e9f1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by robbiejackson** [comment]
I have tested this item :white_check_mark: successfully on ef8201e7c63a4d0aee775aeaee33de63c757e9f1<br /><br />I tested this by just patching the changed files into my 5.1 instance.

I confirmed that with the global config set then the html prefix is restored after you remove it manually.

I also confirmed that an xml feed document was returned correctly whenever a prefix of .feed or a query parameter of format=feed was set in the URL.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by bembelimen** [comment]
@ceford could you please check if the problem still exists?

**Comment by pe7er** [comment]
Thanks @Hackwar !

**Comment by Slava287** [comment]
Hello. When turning on Enforce suffix. Problems appear on multi -language sites. The .html is added, for example: site/de, it becomes site/de.html<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42850">issues.joomla.org/tracker/joomla-cms/42850</a>.</sub>

**Comment by Slava287** [comment]
@Hackwar  Hello. When turning on Enforce suffix. Problems appear on multi -language sites. The .html is added, for example: site/de, it becomes site/de.html   

The problem with the main page

**Comment by brianteeman** [comment]
@Slava287 unable to confirm. Are you using the core Joomla ,htaccess?

**Comment by Slava287** [comment]
@brianteeman

**Comment by Slava287** [comment]
.htaccess

**Comment by Hackwar** [comment]
If you think there is a serious bug, please open a new ticket. Don't respond in closed pull requests, since these are most likely going to be overlooked.

**Comment by Slava287** [comment]
Everything works. I apologize, these were problems on my side.

**Comment by fgsw** [comment]
> these were problems on my side

which?

**Comment by Slava287** [comment]
I had unnecessary redirects to .htaccess

**Comment by fgsw** [comment]
so @brianteeman asked the correct question [comment](https://github.com/joomla/joomla-cms/pull/42850#issuecomment-2621072029)
            

@Slava287 Thanks for the answer. It might help other.

**Pull request merged**